### PR TITLE
Add flexio pwm and flexio spi support for frdm_ke17z and frdm_ke17z512

### PR DESCRIPTION
**Pull request by ankeXiao**
Flexio_pwm: tested `tests/drivers/pwm/pwm_api` `drivers.pwm.ke1xz_pwm_flexio`.
Flexio_spi: tested `tests/drivers/spi/spi_loopback` `drivers.spi.ke1xz_flexio_spi.loopback`.

### CODE REVIEW

File: drivers/spi/spi_mcux_flexio.c
Comment (manuargue) [inline drivers/spi/spi_mcux_flexio.c:108]: use `WAIT_FOR` macro

File: drivers/spi/spi_mcux_flexio.c
Comment (manuargue) [inline drivers/spi/spi_mcux_flexio.c:103]: why does this not apply to other devices?

File: tests/drivers/pwm/pwm_api/testcase.yaml
Comment (manuargue) [inline tests/drivers/pwm/pwm_api/testcase.yaml:16]:
```suggestion
    extra_args: DTC_OVERLAY_FILE="boards/frdm_ke1xz_flexio_pwm.overlay"
```

**Review by manuargue — CHANGES_REQUESTED** [review]

File: tests/drivers/pwm/pwm_api/testcase.yaml
Comment (manuargue) [inline tests/drivers/pwm/pwm_api/testcase.yaml:11]:
```suggestion
  drivers.pwm.ke1xz_pwm_flexio:
```

File: tests/drivers/spi/spi_loopback/testcase.yaml
Comment (manuargue) [inline tests/drivers/spi/spi_loopback/testcase.yaml:185]:
```suggestion
  drivers.spi.ke1xz_flexio_spi.loopback:
```
because the overlay is `frdm_ke1xz_flexio_spi.overlay`

**Review by manuargue — CHANGES_REQUESTED** [review]

File: drivers/spi/spi_mcux_flexio.c
Comment (ankeXiao) [reply drivers/spi/spi_mcux_flexio.c:103]: I don't know if this works on other platforms, I haven't had a chance to test other platforms.

File: drivers/spi/spi_mcux_flexio.c
Comment (ankeXiao) [reply drivers/spi/spi_mcux_flexio.c]: Done

File: tests/drivers/pwm/pwm_api/testcase.yaml
Comment (ankeXiao) [reply tests/drivers/pwm/pwm_api/testcase.yaml]: Updated

File: tests/drivers/spi/spi_loopback/testcase.yaml
Comment (ankeXiao) [reply tests/drivers/spi/spi_loopback/testcase.yaml]: updated

**Comment by ankeXiao** [comment]
Hi @hakehuang , please help connect and test boards.

File: drivers/spi/spi_mcux_flexio.c
Comment (manuargue) [reply drivers/spi/spi_mcux_flexio.c:104]: I don't think is correct to wait on the ISR for `kFLEXIO_SPI_TxBufferEmptyFlag` to be cleared. The interrupt is raised on async transfers and `FLEXIO_SPI_MasterTransferHandleIRQ` is checking for this flag already.
What is the reason you are waiting for this flag to be cleared?

Comment (ankeXiao) [reply drivers/spi/spi_mcux_flexio.c:104]: It doesn't work on my site, the program will get an exception without waiting.
In `FLEXIO_SPI_MasterTransferHandleIRQ`, the status of flexio is indeed detected. However, it is possible that the state has not been stabilized and some data has not been successfully transmitted. So I added the loop to wait for the state to change before calling this function.

image

Comment (ankeXiao) [reply drivers/spi/spi_mcux_flexio.c:104]: And `FLEXIO_SPI_TransferSendTransaction` function calls `FLEXIO_SPI_WriteData` to write data. The TxEmptyFlag needs to be asserted before calling this API. See the description in fsl_flexio.h driver:
image

**Review by manuargue — APPROVED** [review]
LGTM, approving with this remark https://github.com/zephyrproject-rtos/zephyr/pull/76939#discussion_r1713440537 but I'll let other reviewers more familiar with this hw IP comment if further changes are needed

File: tests/drivers/pwm/pwm_api/testcase.yaml
Comment (hakehuang) [inline tests/drivers/pwm/pwm_api/testcase.yaml:20]: why it requires pwm-0 or pwm-1 or pwm-2? do we test all those?

File: tests/drivers/pwm/pwm_api/boards/frdm_ke1xz_flexio_pwm.overlay
Comment (hakehuang) [inline tests/drivers/pwm/pwm_api/boards/frdm_ke1xz_flexio_pwm.overlay:13]: please help t add a comments on line connection

File: tests/drivers/pwm/pwm_api/testcase.yaml
Comment (ankeXiao) [reply tests/drivers/pwm/pwm_api/testcase.yaml:20]: These are required for acquiring pwm device nodes in test_pwm.c.
It is sufficient to test one, and they provide the same configuration.

image

**Review by hakehuang — APPROVED** [review]
board testing pass for pwm and spi_llopback

File: tests/drivers/pwm/pwm_api/boards/frdm_ke1xz_flexio_pwm.overlay
Comment (ankeXiao) [reply tests/drivers/pwm/pwm_api/boards/frdm_ke1xz_flexio_pwm.overlay:13]: No additional connections are required here. This sample outputs a pwm signal from a configured pin when running

**Comment by ankeXiao** [comment]
Hi all, please help review this pull request so we can merge it as soon as possible, thanks!

**Comment by ankeXiao** [comment]
@dleach02 @mmahadevan108 @tbursztyka , could you provide your review so it can move forwards?

**Review by decsny — APPROVED** [review]

File: drivers/spi/spi_mcux_flexio.c
Comment (decsny) [inline drivers/spi/spi_mcux_flexio.c:106]: I would also like to know if this applies to other platforms

Comment (ankeXiao) [reply drivers/spi/spi_mcux_flexio.c:106]: To be on the safe side, I think this macro can be retained and then modified as needed when porting and testing this driver for other platforms in the future.

**Comment by ankeXiao** [comment]
Hello @tbursztyka , could you please provide your review, we want to merge it as soon as possible. Thank you!

**Comment by ankeXiao** [comment]
Can someone help merge this PR?

**Comment by ankeXiao** [comment]
@tbursztyka , please

**Comment by ankeXiao** [comment]
@carlescufi @henrikbrixandersen , could you please help merge this PR into main since it looks like @tbursztyka doesn't work around pwm anymore and this PR has got already the necessary approval. Thanks!

**Comment by tbursztyka** [comment]
@ankeXiao  I never worked on PWM. Your PR should have been split since it brings many features at once (one which is SPI).

Anyway, that PR got already 3 approvals. Looks like it could be merged 
(it's still in DNM status for some reason, though DNM flag is not set)

**Review by tbursztyka — APPROVED** [review]

**Pull request merged**
